File: 623-add-one-row-to-tree/add-one-row-to-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        if (depth == 1) {
            TreeNode* newRoot = new TreeNode(val);
            newRoot->left = root;
            return newRoot;
        }

        queue<TreeNode*> q;
        q.push(root);
        int level = 1;

        while (!q.empty()) {
            int size = q.size();
            if (level == depth - 1) {
                for (int i = 0; i < size; i++) {
                    TreeNode* node = q.front();
                    q.pop();

                    TreeNode* tempLeft = node->left;
                    TreeNode* tempRight = node->right;

                    node->left = new TreeNode(val);
                    node->right = new TreeNode(val);

                    node->left->left = tempLeft;
                    node->right->right = tempRight;
                }
                break;
            }

            for (int i = 0; i < size; i++) {
                TreeNode* node = q.front();
                q.pop();

                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }

            level++;
        }

        return root;
    }
};

File: 623-add-one-row-to-tree/add-one-row-to-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def addOneRow(self, root, val, depth):
        """
        :type root: Optional[TreeNode]
        :type val: int
        :type depth: int
        :rtype: Optional[TreeNode]
        """
        if depth == 1:
            newRoot = TreeNode(val)
            newRoot.left = root
            return newRoot

        queue = [root]
        level = 1

        while queue:
            size = len(queue)
            if level == depth - 1:
                for i in range(size):
                    node = queue.pop(0)

                    tempLeft = node.left
                    tempRight = node.right

                    node.left = TreeNode(val)
                    node.right = TreeNode(val)

                    node.left.left = tempLeft
                    node.right.right = tempRight

                break

            for i in range(size):
                node = queue.pop(0)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            level += 1

        return root
